contrib/profiler/profiler.py

```diff
@@ -188,148 +143,188 @@ def get_network_bytes(filename):
-def analyze_samples(sample_list):
-    count = len(sample_list)
-    min_cpu, min_cpu_idx = sample_list[0].get_cpu_usage(), 0
-    max_cpu, max_cpu_idx = sample_list[0].get_cpu_usage(), 0
-    min_gpu, min_gpu_idx = sample_list[0].get_gpu_usage(), 0
-    max_gpu, max_gpu_idx = sample_list[0].get_gpu_usage(), 0
-    min_mem, min_mem_idx = sample_list[0].get_mem_used(), 0
-    max_mem, max_mem_idx = sample_list[0].get_mem_used(), 0
-    max_read, max_read_idx = sample_list[0].get_read_bytes(), 0
-    sum_cpu, sum_gpu, sum_mem, sum_read = 0, 0, 0, 0
-    cpu_when_gpu_low = list()
-    mem_when_gpu_low = list()
-    disk_read_when_gpu_low = list()
-    for i in range(1, count):
-        if sample_list[i].get_cpu_usage() < min_cpu:
-            min_cpu = sample_list[i].get_cpu_usage()
-            min_cpu_idx = i
-        if sample_list[i].get_cpu_usage() > max_cpu:
-            max_cpu = sample_list[i].get_cpu_usage()
-            max_cpu_idx = i
-        if sample_list[i].get_gpu_usage() < min_gpu:
-            min_gpu = sample_list[i].get_gpu_usage()
-            min_gpu_idx = i
-        if sample_list[i].get_gpu_usage() > max_gpu:
-            max_gpu = sample_list[i].get_gpu_usage()
-            max_gpu_idx = i
-        if sample_list[i].get_mem_used() < min_mem:
-            min_mem = sample_list[i].get_mem_used()
-            min_mem_idx = i
-        if sample_list[i].get_mem_used() > max_mem:
-            max_mem = sample_list[i].get_mem_used()
-            max_mem_idx = i
-        if sample_list[i].get_read_bytes() > max_read:
-            max_read = sample_list[i].get_read_bytes()
-            max_read_idx = i
-
-        if sample_list[i].get_gpu_usage() <= 10:
-            cpu_when_gpu_low.append(sample_list[i].get_cpu_usage())
-            mem_when_gpu_low.append(sample_list[i].get_mem_used())
-            disk_read_when_gpu_low.append(sample_list[i].get_read_bytes())
-
-        sum_cpu += sample_list[i].get_cpu_usage()
-        sum_gpu += sample_list[i].get_gpu_usage()
-        sum_mem += sample_list[i].get_mem_used()
-        sum_read += sample_list[i].get_read_bytes()
-
-    length_gpu_low = len(cpu_when_gpu_low)
-    if length_gpu_low == 0:
-        length_gpu_low = 1
-
-    print('%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f' %
-          (max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-           sum(cpu_when_gpu_low) / length_gpu_low,
-           sum(mem_when_gpu_low) / length_gpu_low,
-           sum(disk_read_when_gpu_low) / length_gpu_low))
-    return [max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-            sum(cpu_when_gpu_low) / length_gpu_low,
-            sum(mem_when_gpu_low) / length_gpu_low,
-            sum(disk_read_when_gpu_low) / length_gpu_low]
+def analyze_samples(sample_list, advise):
+    sample_list = np.array(sample_list, dtype=np.float)
+    used_gpu_num = (sample_list.shape[1] - 7) / 4
```

Comment:
What's the meaning of 7 and 4...? Can you add some comments for the data format of `sample_list`?

Reply:
Maybe change 7 to `GPU_statistics_offset`, change 4 to ` statistics_num_per_GPU`  is more readable.
Also, create a reasonable name for 9, 10 in following code

---
In reply to: [329445955](https://github.com/microsoft/pai/pull/3649#discussion_r329445955) [](ancestors = 329445955)